### PR TITLE
David.benque/queue debugging

### PR DESCRIPTION
**Pull request by dbenque**
In some cases the observability layer was broken due to a label update that was not happening. Initial observations were:
- nodes were continuously added to the queue (labelling require) and that was expected
- when many nodes are added to the queue periodically, sometimes the dequeuing does not happen anymore

After investigation I figured out that the BucketRateLimiter was scheduling the processing time of the item always further in the future when the dequeuing rate was not able to sustain the pressure. Once the processing delay was increased beyond the add frequency, the items would never be processed again.

The problem is that the bucket rate limiter increase does not take into account that the item is already in the queue, it acts globally and end up increasing the delay at every hit.

I manage to reproduce the problem with that small code sample:
https://gist.github.com/dbenque/56c907adc1a1a1f08d16943cd390d7de

The solution is to keep the exponential backup rate limiter at the queue level, and move the BucketRate limiter on the consumption side to ack as a client side protection for the APIServer.

### CODE REVIEW

**Review by antoine-gaillard — APPROVED** [review]

File: internal/kubernetes/observability.go [hunk 1]
```diff
@@ -350,34 +403,34 @@ func (s *DrainoConfigurationObserverImpl) processQueueForNodeUpdates() {
 
 		// func encapsultation to benefit from defer s.queue.Done()
 		func(obj interface{}) {
-			defer s.queueNodeToBeUpdated.Done(obj)
-
 			nodeName := obj.(string)
-			defer s.queueNodeToBeUpdated.Forget(nodeName)
-			// in all case we can forget the node
-			// if needed it will come back at next observability iteration
+			defer s.queueNodeToBeUpdated.Done(obj)
```
Comment (dhenkel92) [inline internal/kubernetes/observability.go:407]: Do we also need to call `.forget(obj)` at some point to reset the failure count of our exponentioal failure rate limiter?

File: internal/kubernetes/observability.go [hunk 2]
```diff
@@ -350,34 +403,34 @@ func (s *DrainoConfigurationObserverImpl) processQueueForNodeUpdates() {
...
+			if s.nodePatchLimiter.TryAccept() {
+				err := s.patchNodeLabels(nodeName)
 				if err != nil {
 					if apierrors.IsNotFound(err) {
-						return nil // the node was deleted, no more need for update.
+						return // the node was deleted, no more need for update.
+					}
+					requeueCount := s.queueNodeToBeUpdated.NumRequeues(nodeName)
+					s.logger.Error("Failed to update label", zap.String("node", nodeName), zap.Int("retry", requeueCount), zap.Error(err))
+					if requeueCount > 10 {
```
Comment (dhenkel92) [inline internal/kubernetes/observability.go:418]: Maybe we want to have another metric about how many are dropped because they seem to be unprocessable?

File: internal/kubernetes/observability.go [hunk 3]
```diff
@@ -107,22 +142,36 @@ type DrainoConfigurationObserverImpl struct {
...
-		configName:          configName,
-		conditions:          conditions,
-		queueNodeToBeUpdated: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
-			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 20*time.Second),
-			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(0.5, 1)}),
-			"nodeUpdater"),
+
+	// We are not adding a BucketRateLimiter to that list because the same nodes are going to be appended periodically if the update fails
+	// Failing nodes will already be in the queue with a retry. Added a BucketRL proved to be a problem here is the client side is not able to dequeue
+	// faster that the add period. In that case the processing time for the item end up being pushed always further in the future and the processing never happens
```
[the remark this answers is not in the document]
Comment (dbenque) [reply internal/kubernetes/observability.go:148]: that really depends on our ability to dequeue in the end... If the add (with rate limit) is done with a period that is faster than our ability to dequeue, then it cannot be processed anymore.
Else I don't understand how the system was blocked for days doing nothing (seen it several times)

File: cmd/draino/draino.go [hunk 4]
```diff
@@ -128,9 +128,9 @@ func main() {
 		// PV/PVC management
 		storageClassesAllowingVolumeDeletion = app.Flag("storage-class-allows-pv-deletion", "Storage class for which persistent volume (and associated claim) deletion is allowed. May be specified multiple times.").PlaceHolder("storageClassName").Strings()
 
-		configName           = app.Flag("config-name", "Name of the draino configuration").Required().String()
-		resetScopeAnnotation = app.Flag("reset-config-annotations", "Reset the scope annotation on the nodes").Bool()
-		scopeAnalysisPeriod  = app.Flag("scope-analysis-period", "Period to run the scope analysis and generate metric").Default((5 * time.Minute).String()).Duration()
+		configName          = app.Flag("config-name", "Name of the draino configuration").Required().String()
+		resetScopeLabel     = app.Flag("reset-config-labels", "Reset the scope label on the nodes").Bool()
```
[the remark this answers is not in the document]
Comment (dbenque) [reply cmd/draino/draino.go:132]: @adrienjt wanted us to move the value from annotation to label so that it was queryable via labelSelector. When @antoine-gaillard did the implementation he forgot to change some comments and flag name to align the implementation with the vocabulary. I am also fixing this in this PR.

File: internal/kubernetes/observability.go (hunk 1 above)
Comment (dbenque) [reply internal/kubernetes/observability.go:407]: I do it later only in case we have reach a high number of retries.

Comment (dhenkel92) [reply internal/kubernetes/observability.go:407]: But shouldn't the failure count be reseted after it was successfully processed?
As far as I understand it will work like this right now:
```
- add "a" (failure count 0)
- process "a" -> failed -> requeue (failure count 1)
- process "a" -> failed -> requeue (failure count 2)
- process "a" -> success
- add "a" (failure count 2)
```

File: cmd/draino/draino.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhenkel92) [reply cmd/draino/draino.go:132]: 😄 okay, I see

File: internal/kubernetes/observability.go (hunk 1 above)
Comment (dbenque) [reply internal/kubernetes/observability.go:407]: indeed, the node should not come back because it is labelled now, but you hare right, it should be forgotten in case of success.
Fixing it

File: internal/kubernetes/observability.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbenque) [reply internal/kubernetes/observability.go:148]: you are right, it will happen at some point... takes a very long time but happens. Fixing the comment.

File: internal/kubernetes/observability.go (hunk 2 above)
Comment (dbenque) [reply internal/kubernetes/observability.go:418]: They are no really dropped, because the observability layer is looping periodically on all the object. We just have to wait on the next round.
I don't know if it will be actionable. What would be actionable is that we see that the same node is being dropped again, again and again. So that would deserve a metrics with node name as tag.
I think digging into the logs if the metrics `MeasureNodeLabelPatchFailed` got multiple hit (x9) will be enough.
Basically we can alert and investigate in case the `MeasureNodeLabelPatchFailed` metrics is continuously increasing (at the same rate) over 15 minutes.

Comment (dhenkel92) [reply internal/kubernetes/observability.go:412]: Sounds reasonable 👍

**Review by dhenkel92 — APPROVED** [review]
